rs/replicated_state/src/canister_state/queues.rs

```diff
@@ -393,18 +410,32 @@ impl CanisterQueues {
             }
             RequestOrResponse::Response(ref response) => {
                 match self.canister_queues.get_mut(&sender) {
-                    Some((queue, _)) if queue.check_has_reserved_response_slot().is_ok() => queue,
+                    Some((queue, _)) if queue.check_has_reserved_response_slot().is_ok() => {
+                        // Check against duplicate responses.
+                        if !self
+                            .callbacks_with_enqueued_response
+                            .insert(response.originator_reply_callback)
+                        {
+                            debug_assert_eq!(Ok(()), self.test_invariants());
+                            // This is a critical error for guaranteed responses.
+                            if response.deadline == NO_DEADLINE {
+                                return Err((
+                                    non_matching_response("Duplicate response", response),
+                                    msg,
```

Comment:
Just to double check: returning an error here will be handled upstream by raising a critical error but _not_ crashing right?

Reply:
There is no exception for just this error type so it doesn't add any panics that were not there before.